File: src/Functional/Scala/Composite.scala
sealed trait Node {
  def size: Int
}

final case class FileLeaf(bytes: Int) extends Node {
  override def size: Int = bytes
}

final case class FolderComposite(children: List[Node]) extends Node {
  override def size: Int = children.map(_.size).sum
}

object Composite {
  def main(args: Array[String]): Unit = {
    val readme: Node = FileLeaf(2)
    val docs: Node = FolderComposite(List(FileLeaf(3), FileLeaf(5)))
    val root: Node = FolderComposite(List(readme, docs))

    println(s"leaf=${readme.size}")
    println(s"docs=${docs.size}")
    println(s"root=${root.size}")
  }
}

File: src/Functional/Clojure/composite.clj
(defn file-node [bytes]
  {:kind :file :bytes bytes})

(defn folder-node [& children]
  {:kind :folder :children (vec children)})

(defn node-size [node]
  (if (= :file (:kind node))
    (:bytes node)
    (reduce + (map node-size (:children node)))))

(let [readme (file-node 2)
      docs (folder-node (file-node 3) (file-node 5))
      root (folder-node readme docs)]
  (println (str "leaf=" (node-size readme)))
  (println (str "docs=" (node-size docs)))
  (println (str "root=" (node-size root))))
